[Merged by Bors] - feat(linear_algebra/alternating): Add dom_coprod

eric-wieser opened the pull request:
This implements a variant of the multiplication defined in the second half of [Proposition 22.24 of "Notes on Differential Geometry and Lie Groups" (Jean Gallier)](https://www.cis.upenn.edu/~cis610/diffgeom-n.pdf):

![image](https://user-images.githubusercontent.com/425260/104042315-3dfe7380-51d2-11eb-9b3a-bbbb52d180f0.png)

---
<!--
put comments you want to keep out of the PR commit here.
If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
-->

Before this gets merged I'll want to squash the commits to remove @zhangir-azerbayev from the history, as none of the code changed in this PR is theirs.

<details>

<summary> Resolved Dependencies </summary>

- [x] depends on: #5187
- [x] depends on: #5260
- [x] depends on: #5266
- [x] depends on: #5270
- [x] depends on: #5279
- [x] depends on: #5368
- [x] depends on: #5365
- [x] depends on: #5364
- [x] depends on: #5372
- [x] depends on: #5374
- [x] depends on: #5301
- [x] depends on: #5396
- [x] depends on: #5317
- [x] depends on: #5430
- [x] depends on: #5509

</details>

- [x] depends on: #6186

eric-wieser commented on src/group_theory/perm/shuffle.lean:
```diff
@@ -0,0 +1,80 @@
+import group_theory.perm.sign
```
This file is junk, I'll delete it when I next push.

eric-wieser commented on src/linear_algebra/alternating.lean:
```diff
@@ -37,6 +43,57 @@ using `map_swap` as a definition, and does not require `has_neg N`.
+  [fintype α] [fintype β] [group α] [group β] [add_comm_monoid γ]
+  (f : α →* β) (h : function.injective f) {_ : decidable_pred (λ (b : β), b ∈ f.range)} (g : β → γ) :
+  ∑ x in finset.filter (λ (b : β), b ∈ f.range) finset.univ, g x = ∑ a, g (f a) :=
+finset.sum_filter_univ_exists_eq f h g
```
[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there.20code.20for.20X.3F/topic/Summing.20over.20.60monoid_hom.2Erange.60/near/222093999)

jcommelin commented on src/linear_algebra/alternating.lean:
```diff
@@ -37,6 +43,57 @@ using `map_swap` as a definition, and does not require `has_neg N`.
+open interactive
+
+/-- Focus on the first `n` goals. -/
+meta def first_n_goals : parse small_nat → itactic → tactic unit
```
do you think `work_on_goal` could be generalised to include this use case?

eric-wieser replied:
Not really, because `work_on_goal n` has semantics "work on the goal, and them emit `old_goals[:n] ++ new_goals ++ old_goals[n + 1:]`. If you're working on multiple goals, there's no sensible choice for where to insert the new goals.

eric-wieser commented on src/linear_algebra/alternating.lean:
```diff
@@ -434,3 +444,292 @@ lemma comp_multilinear_map_alternatization (g : N' →ₗ[R] N'₂)
+        cases hj : σ⁻¹ j;
+        rw perm.inv_eq_iff_eq at hi hj;
+        propagate_tags' { substs hi hj, }, },
+    case [sum.inl sum.inr : i' j', sum.inr sum.inl : i' j'] {
```
@gebner, this is where I needed `propagate_tags` from leanprover-community/lean#540. Can you think of a better solution? Should `substs` just call `propagate_tags` itself?

gebner replied:
Good lord! I think `propagate_tags` is the least of this proof's problems. Is there any way it can be split up into more manageable and maybe even independently useful chunks?

But yes, `substs` should call `propagate_tags` itself.

gebner replied:
Sorry for the first impressions.  I don't find it particularly readable when every line is either `all_goals` or `work_on_goal`.  While I value avoiding duplication, I think it would be better to refactor `all_goals { a }, work_on_goal 0 { b }, work_on_goal 1 { b' }, all_goals { c }` into `{ a, b, c }, { a, b', c }` here.

I would also introduce a definition for the summand of `dom_coprod_fun`.  This is an unwieldy expression, and you can probably extract two of the big blocks in the proof here.

eric-wieser replied:
> I would also introduce a definition for the summand of dom_coprod_fun. This is an unwieldy expression, and you can probably extract two of the big blocks in the proof here.

I've done this as `dom_coprod.summand` and eliminated `dom_coprod_fun` - good idea.

gebner commented on src/linear_algebra/alternating.lean:
```diff
@@ -434,3 +444,292 @@ lemma comp_multilinear_map_alternatization (g : N' →ₗ[R] N'₂)
+  [comm_semiring R]
+  [add_comm_group N₁] [semimodule R N₁]
+  [add_comm_group N₂] [semimodule R N₂]
+  [add_comm_monoid M] [semimodule R M]
```
I would use `variables` for this.

eric-wieser replied:
I didn't do this because there is already a global `R` variable in the file with an associated `semiring`. It's frustrating that `def ... {R : Type*}` can shadow variables, but `section` `variables {R : Type*}` `def ...` can't.

I guess I'll just introduce an `R'` and use that instead...

gebner commented on src/linear_algebra/alternating.lean:
```diff
@@ -434,3 +444,292 @@ lemma comp_multilinear_map_alternatization (g : N' →ₗ[R] N'₂)
+* the filtering of $[1, m+n]$ to shuffles with an isomorphic quotient
+* the additions in the subscripts of $\sigma$ with a index of type `sum`
+-/
+@[simps]
```
```suggestion
```
As far as I understand, `dom_coprod_fun` is an implementation detail here, and you don't want `dom_coprod` to unfold into `dom_coprod_fun`.

jcommelin commented on src/linear_algebra/alternating.lean:
```diff
@@ -434,3 +443,280 @@ lemma comp_multilinear_map_alternatization (g : N' →ₗ[R] N'₂)
+Here, we generalize this by replacing:
+* the product in the sum with a tensor product
+* the filtering of $[1, m+n]$ to shuffles with an isomorphic quotient
+* the additions in the subscripts of $\sigma$ with a index of type `sum`
```
```suggestion
* the additions in the subscripts of $\sigma$ with an index of type `sum`
```

jcommelin commented on src/linear_algebra/alternating.lean:
```diff
@@ -434,3 +443,280 @@ lemma comp_multilinear_map_alternatization (g : N' →ₗ[R] N'₂)
+      (λ σ _, finset.mem_univ _)
+      (λ σ _, equiv.perm.mod_sum_congr.swap_smul_involutive i j σ),
+  end,
+  ..(∑ σ : perm.mod_sum_congr ιa ιb, dom_coprod.summand a b σ)}
```
```suggestion
  ..(∑ σ : perm.mod_sum_congr ιa ιb, dom_coprod.summand a b σ) }
```

jcommelin reviewed (approved):
Thanks :tada:

bors d+

eric-wieser commented on docs/references.bib:
```diff
@@ -281,6 +281,13 @@ @InProceedings{   fuerer-lochbihler-schneider-traytel2020
   bibsource     = {dblp computer science bibliography, https://dblp.org}
 }
 
+@Inproceedings{   Gallier2011Notes,
```
This was what "semantic scholar" suggested, but I have no idea what the right choice is.

bryangingechen replied:
LGTM as far as `doc-gen` is concerned.

eric-wieser commented on src/linear_algebra/alternating.lean:
```diff
@@ -434,3 +443,280 @@ lemma comp_multilinear_map_alternatization (g : N' →ₗ[R] N'₂)
+
+/-- Like `multilinear_map.dom_coprod`, but ensures the result is also alternating.
+
+Note that this is usually defined over integer indices `ιa = fin n` and `ιb = fin m`, as
```
```suggestion
Note that this is usually defined (for instance, as used in Proposition 22.24 in [Gallier2011Notes])
over integer indices `ιa = fin n` and `ιb = fin m`, as
```

bryangingechen commented on docs/references.bib:
```diff
@@ -281,6 +281,13 @@ @InProceedings{   fuerer-lochbihler-schneider-traytel2020
+  title         = {Notes on Differential Geometry and Lie Groups},
+  author        = {J. Gallier and J. Quaintance},
+  year          = {2011},
+  url           = {http://www.cis.upenn.edu/~cis610/diffgeom-n.pdf}
```
This seems to redirect to the `https` link immediately:
```suggestion
  url           = {https://www.cis.upenn.edu/~cis610/diffgeom-n.pdf}
```

eric-wieser committed:
feat(linear_algebra/alternating): Add dom_coprod (#5269)

This implements a variant of the multiplication defined in the second half of [Proposition 22.24 of "Notes on Differential Geometry and Lie Groups" (Jean Gallier)](https://www.cis.upenn.edu/~cis610/diffgeom-n.pdf):

![image](https://user-images.githubusercontent.com/425260/104042315-3dfe7380-51d2-11eb-9b3a-bbbb52d180f0.png)
